modyn/trainer_server/internal/trainer/remote_downsamplers/abstract_matrix_downsampling_strategy.py

```diff
@@ -0,0 +1,122 @@
+        # yields the following downsampled class sizes [5, 25, 15] while without balance something like [0, 45, 0] can
+        # happen
+        self.balance = params_from_selector["balance"]
+        if self.balance:
+            # Selection happens class by class. These data structures are used to store the selection results
+            self.already_selected_samples: list[int] = []
+            self.already_selected_weights = torch.tensor([]).float()
+            self.requires_data_label_by_label = True
+        else:
+            self.requires_data_label_by_label = False
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        assert embedding is not None
+        assert self.matrix_content is not None
+
+        if self.matrix_content == MatrixContent.GRADIENTS:
+            new_elements = self._compute_gradients(forward_output, target, embedding)
+        elif self.matrix_content == MatrixContent.EMBEDDINGS:
+            new_elements = embedding.detach().cpu()
+        else:
+            raise AssertionError("The required content does not exits.")
+
+        self.matrix_elements.append(new_elements)
+        # keep the mapping index<->sample_id
+        self.index_sampleid_map += sample_ids
+
+    def _compute_gradients(
+        self, forward_output: torch.Tensor, target: torch.Tensor, embedding: torch.Tensor
+    ) -> torch.Tensor:
+        loss = self.criterion(forward_output, target).mean()
+        embedding_dim = embedding.shape[1]
+        num_classes = forward_output.shape[1]
+        batch_num = target.shape[0]
+        # compute the gradient for each element provided
+        with torch.no_grad():
+            bias_parameters_grads = torch.autograd.grad(loss, forward_output)[0]
+            weight_parameters_grads = embedding.view(batch_num, 1, embedding_dim).repeat(
+                1, num_classes, 1
+            ) * bias_parameters_grads.view(batch_num, num_classes, 1).repeat(1, 1, embedding_dim)
+            gradients = torch.cat([bias_parameters_grads, weight_parameters_grads.flatten(1)], dim=1).cpu().numpy()
+        return gradients
+
+    def inform_end_of_current_label(self) -> None:
+        assert self.balance
+        selected_samples, selected_weights = self._select_from_matrix()
+        self.already_selected_samples += selected_samples
+        self.already_selected_weights = torch.cat((self.already_selected_weights, selected_weights))
+        self.matrix_elements = []
+        self.index_sampleid_map: list[int] = []
+
+    def select_points(self) -> tuple[list[int], torch.Tensor]:
+        if self.balance:
+            ids, weights = self.already_selected_samples, self.already_selected_weights
+        else:
+            ids, weights = self._select_from_matrix()
+
+        return _shuffle_list_and_tensor(ids, weights)
```

Comment:
Could you briefly explain why we shuffle the selection again?

Reply:
Training is done class by class if we do not shuffle with the balance version (first $N_0$ points of class 0, then $N_1$ points of class 1 and so on). In non-balance, it is less important, but I think it is still useful (otherwise, the training follows the order of importance of the selection, i.e. the first batches have all the most important points and then progressively less and less important points come in).

Reply:
Thank you!